nodejs/lesson02/README.md

```diff
@@ -15,26 +10,134 @@ Learn how to:
+  const helloStr = `Hello, ${helloTo}!`;
+  span.log({
+    event: "string-format",
+    value: helloStr,
+  });
+  span.finish();
+  return helloStr;
+};
+
+const printString = (rootSpan, helloStr) => {
+  const span = tracer.startSpan("consoleLog", { childOf: rootSpan });
+  console.log(helloStr);
+  span.log({ event: "print-string" });
+  span.finish();
+};
 ```
 
-You should see something below on the console for the client app:
+If we think of the trace as a directed acyclic graph where nodes are the spans and edges are the causal relationships between them, then the `childOf` option is used to create one such edge between `span` and `rootSpan`. In the API the edges are represented by `SpanReference` type that consists of a `SpanContext` and a label. The `SpanContext` represents an immutable, thread-safe portion of the span that can be used to establish references or to propagate it over the wire. The label, or `ReferenceType`, describes the nature of the relationship. `ChildOf` relationship means that the `rootSpan` has a logical dependency on the child `span` before `rootSpan` can complete its operation. Another standard reference type in OpenTracing is `FollowsFrom`, which means the `rootSpan` is the ancestor in the DAG, but it does not depend on the completion of the child span, for example if the child represents a best-effort, fire-and-forget cache write.
+
+If we modify the `formatString` and `printString` functions accordingly and run the app, we'll see that all reported spans now belong to the same trace:
 
 ```
-INFO  Reporting span d6c674cb77b1ba9c:933a7b0d05ed1b48:d6c674cb77b1ba9c:1
-Hello, Peter!
-INFO  Reporting span d6c674cb77b1ba9c:263fe6d3cb907321:d6c674cb77b1ba9c:1
-INFO  Reporting span d6c674cb77b1ba9c:d6c674cb77b1ba9c:0:1
-```
\ No newline at end of file
+node lesson02/exercise/hello.js Kara
+INFO  Initializing Jaeger Tracer with CompositeReporter and ConstSampler
+INFO  Reporting span f807cdcd1b44f817:9cb1e0041868bfcd:f807cdcd1b44f817:1
+Hello, Kara!
+INFO  Reporting span f807cdcd1b44f817:6d38799352b613ec:f807cdcd1b44f817:1
+INFO  Reporting span f807cdcd1b44f817:f807cdcd1b44f817:0:1
+```
+
+We can also see that the first hexadecimal segment of the output for all three spans is `f807cdcd1b44f817`, which is the Jaeger trace ID of the root span. Additionally, the first two reported spans display the trace ID of the root span in the 3rd position, instead of `0`. The root span is reported last because it is the last one to finish.
+
+If we find this trace in the UI, it will show a proper parent-child relationship between the spans.
+
+### Propagate the in-process context
+
+You may have noticed one unpleasant side effect of our recent changes - we had to pass the Span object as the first argument to each function. JavaScript does not support the notion of thread-local variables,
+so in order to link the individual spans together we _do need to pass something_.
```

Comment:
Yeah, it's tricky in Node. There are two options:
  1. create some simple Context class (similar to Go) and pass it to all functions (this is what we do at Uber)
  2. use continuation-local-storage to store the span (https://github.com/openzipkin/zipkin-js does that)